core/command_parser.h

```diff
@@ -0,0 +1,246 @@
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2020 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2020 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef COMMAND_PARSER_H
+#define COMMAND_PARSER_H
+
+#include "core/object/reference.h"
+#include "core/templates/ordered_hash_map.h"
+
+class CommandOption : public Reference {
+	GDCLASS(CommandOption, Reference);
+
+	struct ArgumentChecker;
+	struct FunctionChecker;
+	struct CallableChecker;
+
+	// Names for the command. e.g --help or -h.
+	PackedStringArray _names;
+	// List of allowed values for each argument, empty if any value is allowed.
+	PackedStringArray _default_args;
+	// List of allowed values for each argument, empty if any value is allowed.
+	PackedStringArray _allowed_values;
+	// Command description that will be displayed in help.
+	String _description;
+	// Command category, commands sharing the same category are grouped together in the help text.
+	String _category;
+	// Name for the command arguments that will be displayed in help text.
+	String _metavar = RTR("<arg>");
+	// Make the command visible in help.
+	bool _show_in_help = true;
```

Comment:
Could be useful if you want to have help several messages for different options. [Here](https://www.boost.org/doc/libs/1_76_0/doc/html/program_options/howto.html#id-1.3.32.6.6) is how it works in Boost. But let me know if this could be an overkill.

Reply:
Makes sense, but it may be worth to rename this to `hidden`, because technically `CommandLineOption` class is not "aware" of the parser's ability to generate help text, this is parser's responsibility, rather (talking about some level of abstraction).

Reply:
Agree, thanks!